openmls/src/group/tests/test_proposal_validation.rs

```diff
@@ -0,0 +1,190 @@
+        proposal_store.add(
+            QueuedProposal::from_mls_plaintext(ciphersuite, backend, proposal.to_owned())
+                .expect("Could not create QueuedProposal from MlsPlaintext"),
+        );
+    }
+    proposal_store
+}
+
+/// ValSem100:
+/// Add Proposal:
+/// Identity in proposals must be unique among proposals
+#[apply(ciphersuites_and_backends)]
+fn test_valsem100(ciphersuite: &'static Ciphersuite, backend: &impl OpenMlsCryptoProvider) {
+    let (alice_credential_bundle, alice_key_package_bundle) =
+        generate_credential_bundle_and_key_package_bundle("Alice".into(), ciphersuite, backend);
+
+    // 0. Create Bob and Charlie with the same identity
+    let (_bob_credential_bundle, bob_key_package_bundle) =
+        generate_credential_bundle_and_key_package_bundle("42".into(), ciphersuite, backend);
+    let bob_key_package = bob_key_package_bundle.key_package().clone();
+
+    let (_charlie_credential_bundle, charlie_key_package_bundle) =
+        generate_credential_bundle_and_key_package_bundle("42".into(), ciphersuite, backend);
+    let charlie_key_package = charlie_key_package_bundle.key_package().clone();
+
+    // 1. Alice creates a group
+    let group_aad = b"Alice's Friends";
+    let framing_parameters = FramingParameters::new(group_aad, WireFormat::MlsCiphertext);
+    let alice_group = CoreGroup::builder(GroupId::random(backend), alice_key_package_bundle)
+        .build(backend)
+        .expect("Error creating group.");
+
+    // 2. Alice creates a proposal to add Bob
+    let bob_add_proposal = alice_group
+        .create_add_proposal(
+            framing_parameters,
+            &alice_credential_bundle,
+            bob_key_package,
+            backend,
+        )
+        .expect("Could not create proposal to add Bob.");
+
+    // 3. Alice creates a proposal to add Charlie
+    let charlie_add_proposal = alice_group
+        .create_add_proposal(
+            framing_parameters,
+            &alice_credential_bundle,
+            charlie_key_package,
+            backend,
+        )
+        .expect("Could not create proposal to add Charlie.");
+
+    // 4. Alice queues these proposals
+    let proposal_store = generate_proposal_store(
+        &[bob_add_proposal, charlie_add_proposal],
+        ciphersuite,
+        backend,
+    );
+
+    // 5. Alice tries to generate a commit message
+    let params = CreateCommitParams::builder()
+        .framing_parameters(framing_parameters)
+        .credential_bundle(&alice_credential_bundle)
+        .proposal_store(&proposal_store)
+        .build();
+    let err = alice_group
+        .create_commit(params, backend)
+        .expect_err("Created commit when the proposals have a duplicate identity!");
+    assert_eq!(
+        err,
+        CoreGroupError::ProposalValidationError(
+            ProposalValidationError::DuplicateIdentityAddProposal
+        )
+    );
+
+    // TODO #525: Also test that we are correctly validating incoming proposals.
```

Comment:
In the other validation tests, each test consists of two tests, a negative one (which is already here), where we check that the validation fails as specified and a positive one, where the validation goes through. It would be great, if you could add a simple positive case as well. Or is that why you flagged #525 here and mark the test as partly implemented? Same goes for the other test below.

Reply:
Good catch, I forgot to add a positive test, I will add one.

Reply:
The TODO is for alice receiving a malformed proposal message from another group member bob. I left that as a TODO for now because to generate the malformed proposal, we need a test version of `create_commit` that does not run the validation checks. Alternatively, if we move the validation checks to `ProposalStore` as you suggested, we wouldn't have to do that.